### PR TITLE
Fix some query part expressions not being cloned.

### PR DESCRIPTION
**Pull request by ndm2**
Some parts have an array structure that stores possible expressions in a
nested fashion, these require some specific logic in order to properly
clone them.

### CODE REVIEW

File: src/Database/Query.php [hunk 1]
```diff
@@ -2327,9 +2330,41 @@ public function __clone()
...
             if (is_array($part)) {
                 foreach ($part as $i => $piece) {
-                    if ($piece instanceof ExpressionInterface) {
-                        /** @psalm-suppress PossiblyUndefinedMethod */
-                        $this->_parts[$name][$i] = clone $piece;
+                    switch ($name) {
+                        case 'join':
+                            if ($piece['table'] instanceof ExpressionInterface) {
+                                /** @psalm-suppress PossiblyUndefinedMethod */
+                                $this->_parts[$name][$i]['table'] = clone $piece['table'];
```
Comment (othercorey) [inline src/Database/Query.php]: What method is undefined?

Comment (ndm2) [reply src/Database/Query.php]: `Cake\Database\ExpressionInterface::offsetSet`

https://github.com/ndm2/cakephp/runs/3984618896?check_suite_focus=true

File: src/Database/Query.php [hunk 2]
```diff
@@ -2327,9 +2330,41 @@ public function __clone()
             }
             if (is_array($part)) {
                 foreach ($part as $i => $piece) {
-                    if ($piece instanceof ExpressionInterface) {
-                        /** @psalm-suppress PossiblyUndefinedMethod */
-                        $this->_parts[$name][$i] = clone $piece;
+                    switch ($name) {
```
Comment (othercorey) [inline src/Database/Query.php]: Could we iterate on all keys and check?

Comment (ndm2) [reply src/Database/Query.php]: Not sure what exactly you mean?

Comment (ndm2) [reply src/Database/Query.php]: Oh wait, you mean irrespective of the clause name iterate all `$piece` if it's an array?

Comment (othercorey) [reply src/Database/Query.php]: yea, just clone anything that's an expression.

**Comment by othercorey** [comment]
Looks ok to me unless someone knows a reason why not to clone these.

**Pull request merged**